File: Array/zhangjianan/20200618队列的最大值/MaxQueue.java
package com.example.testapplication;

import java.util.Deque;
import java.util.LinkedList;
import java.util.Queue;

public class MaxQueue {
    Queue<Integer> queue=new LinkedList<Integer>();
    Deque<Integer> deque=new LinkedList<Integer>();

    public MaxQueue() {
    }

    public int max_value() {
        return deque.size()>0?deque.peek():-1;
    }

    public void push_back(int value) {
        queue.offer(value);
        while (deque.size() > 0 && deque.peekLast() < value) {
            deque.pollLast();
        }
        deque.offerLast(value);
    }

    public int pop_front() {
        int temp=queue.size()>0?queue.poll():-1;
        if (deque.size()>0&&temp==deque.peek()) {
            deque.poll();
        }
        return temp;
    }
}

File: Array/zhangjianan/20200618队列的最大值/MaxQueue.kt
package com.example.testapplication

import java.util.*

class MaxQueue {
    var queue: Queue<Int> = LinkedList()
    var deque: Deque<Int> = LinkedList()
    fun max_value(): Int {
        return if (deque.size > 0) deque.peek() else -1
    }

    fun push_back(value: Int) {
        queue.offer(value)
        while (deque.size > 0 && deque.peekLast() < value) {
            deque.pollLast()
        }
        deque.offerLast(value)
    }

    fun pop_front(): Int {
        val temp = if (queue.size > 0) queue.poll() else -1
        if (deque.size > 0 && temp == deque.peek()) {
            deque.poll()
        }
        return temp
    }

}